refactor(linter/plugins): use `Record` types

overlookmotel opened the pull request:
Pure refactor. We were using a mix of `{ [key: string]: T }` and `Record<string, T>`. Standardize on `Record`.

overlookmotel commented:
* **#15379** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15379?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15379?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15379).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 reviewed (approved):
enable the lint rule for this? 😉

overlookmotel committed:
refactor(linter/plugins): use `Record` types (#15379)

Pure refactor. We were using a mix of `{ [key: string]: T }` and `Record<string, T>`. Standardize on `Record`.

overlookmotel commented:
> enable the lint rule for this? 😉

Ooh yes. Which one?
